docs/ronn/knox.1.ronn
[comment]: # ( -*- Mode: markdown; -*- )
knox(1) -- run a non-interactive Kno script
=========================================

## SYNOPSIS

`knox` [config=val]... <source> [config=val|arg]...

## DESCRIPTION

This invokes the Kno interpreter to load and run a program (indicated
by *source*) after processing any provided config entries. Config
entries have the form **setting**=*value* and may trigger the
execution of procedural *config handlers*.

If *source* is a filename, the file is opened and expressions are read
and evaluated; if *source* is the string '-', expressions are read and
evaluated from the standard input. If those expressions define a
function `main`, `main` is called on the command line *args*
(excluding config entries).

**exec modules**: if *source* starts with a colon (`:`), it refers to
a Scheme module which will be loaded into the `knox` session using the
normal search paths as modified by the config entries. If the loaded
module exports a `main` procedure, that procedure is applied to the
command line *args* (excluding config entries).

Some config entries may generate useful side-effects defined by either
the Kno core or loaded modules. Some useful built-in config
settings are:

* `CONFIG=`*config file*:
  loads a config file and processes it's entries.

* `MODULES=`*module name*:
  loads a Kno module into the image and the default 
  environment.

* `LOGLEVEL=`*level*:
  specifies the default loglevel for notifications
  where level is either a positive integer or symbolic name (like
  `NOTICE` or `DEBUG` or `INFO`).

* `LOADFILE=`*path*:
  loads the file *path* during startup.

* `LOADPATH=`*path*:
  specifies a search path for modules, as a colon
  separated list of patterns where the '%' character is replaced by
  the module name. For example `/opt/mods/%/module.scm:/opt/mods/%:`
  causes the system to look for a module `foo/x` in
  `/opt/mods/foo/x/module.scm` and `/opt/mods/foo/x.scm`. Paths can be
  relative to the current directory by starting with '.' and can also
  include ~home directory references.

## DIAGNOSTICS

Returns 0 when exiting normally, 1 otherwise. It also displays a short
preamble on starting and reports its time and memory usage on exit.

## IMPLEMENTING SCRIPTS

On Unix systems, `knox` can be used with a *shebang* in executable
script files, so that Kno source file is executable and starts with:

`#!/usr/bin/knox LOGLEVEL=5`

or

`#!/usr/bin/env knox LOGLEVEL=5`

knox will read the file after processing the arguments in the
initial line (such as loglevel).


## FILES

Unlike `knoc`, the interactive Kno console, `knox` does not read the
user's `.knoconfig` file, though this can be specified by the
**CONFIG** config setting.

## ENVIRONMENT

As with most Kno programs, config variables for `knox` can be
set through environment variables prefixed with `KNO`, so `KNO_LOGLEVEL`
will set the default loglevel.

# AUTHOR

Ken Haase <kh@beingmeta.com>

## SEE ALSO

knoc(1), kno_config(8)



debian/changelog
framerd (3.9.1-13) unstable; urgency=medium

  * Fix packaging descriptions
  * git patch version framerd-3.9.1-31-g2af7e18

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 11:11:23 -0400

framerd (3.9.1-12) unstable; urgency=medium

  * Split out framerd into framerd-core and framerd-builtin (which provides framerd-modules)
  * git patch version framerd-3.9.1-29-ge3a33fa

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 10:58:54 -0400

framerd (3.9.1-11) unstable; urgency=low

  * More adjustments to libmysqlclient dependencies
  * git patch version framerd-3.9.1-25-g880479a

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 10:12:25 +0000

framerd (3.9.1-10) unstable; urgency=low

  * Fix libmysqlclient dependency
  * Save alternate booksystem files in config directory
  * git patch version framerd-3.9.1-23-g4def5c8

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 09:17:44 +0000

framerd (3.9.1-9) unstable; urgency=low

  * Add libzip2 dependency to packaging
  * Fix FramerD systemd configuration
  * git patch version framerd-3.9.1-18

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 14:47:16 +0000

framerd (3.9.1-8) unstable; urgency=low

  * Add optional arg to empty-string? to treat non-breaking spaces as content
  * git patch version framerd-3.9.1-15-gff06aeb

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 03:51:54 +0000

framerd (3.9.1-7) unstable; urgency=low

  * Fix bug in optimize code
  * git patch version framerd-3.9.1-12-g99c9b27

 -- Ken Haase <kh@beingmeta.com>  Sat, 30 Apr 2016 22:51:01 +0000

framerd (3.9.1-6) unstable; urgency=medium

  * Repackage for alternate libmagickwand under xenial
  * git patch version framerd-3.9.1-9-gad2c5fc

 -- Ken Haase <kh@beingmeta.com>  Fri, 29 Apr 2016 13:16:58 -0400

framerd (3.9.1-5) unstable; urgency=medium

  * Improve error capture and handling including interrupts
  * Add fdstatus and fdlog commands
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Thu, 31 Mar 2016 07:01:55 -0400

framerd (3.9.0-4) unstable; urgency=medium

  * Fix build flags to remove needless dependencies
  * git patch version framerd-3.9-15-ge4f2745

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 17:35:42 -0400

framerd (3.9.0-3) unstable; urgency=low

  * Fix opcode handling bugs
  * git patch version framerd-3.9-6-gc9661e4

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 11:14:47 -0400

framerd (3.9.0-2) unstable; urgency=low

  * Fix git tag and executable heralds
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 21:13:27 -0400

framerd (3.9.0-1) unstable; urgency=low

  * Fix systemd scripts
  * Added fdstatus script
  * Fix pool-load to set error context
  * Fixes to opcode bugs
  * Bumped minor release number

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 16:55:24 -0400

framerd (3.8.4-63) unstable; urgency=low

  * Fix libbson/libmongoc dependencies
  * git patch verison framerd-3.8.4-177-gdb5c379

 -- Ken Haase <kh@beingmeta.com>  Sun, 27 Mar 2016 00:31:54 +0000

framerd (3.8.4-62) unstable; urgency=low

  * Fix tidy5 dependencies
  * git patch version framerd-3.8.4-176-g33eb981

 -- Ken Haase <kh@beingmeta.com>  Sat, 26 Mar 2016 14:25:11 +0000

framerd (3.8.4-61) unstable; urgency=low

  * Add smarter immediate type checking
  * Fix some opcode/optimize bugs
  * git patch version framerd-3.8.4-173-gdc93e81

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 21:33:36 -0400

framerd (3.8.4-60) unstable; urgency=low

  * Fixes to xmlparsing
  * Fix some applytest error cases (whoops)
  * Check irritant pointer when calling fd_err
  * Improve backtrace CSS
  * git patch version framerd-3.8.4-168-g8a0394e

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 18:21:37 -0400

framerd (3.8.4-59) unstable; urgency=low

  * Add handling of extpool lockfn
  * make DOM cleanup take more initiative
  * git patch version framerd-3.8.4-156-g43cb8ea

 -- Ken Haase <kh@beingmeta.com>  Sat, 19 Mar 2016 19:51:14 +0000

framerd (3.8.4-58) unstable; urgency=low

  * Add set-procedure-args!
  * Add functions for augmenting errors in ONERROR handlers
  * Fixes to optimize module
  * Wrap primitives execution in call contours to convert setmped errors
    to return errors
  * git patch version framerd-3.8.4-153-gcd07c91

 -- Ken Haase <kh@beingmeta.com>  Mon, 14 Mar 2016 18:26:35 -0400

framerd (3.8.4-57) unstable; urgency=low

  * Pass priority on to logger from logif+
  * git patch version framerd-3.8.4-147-gfb67432

 -- Ken Haase <kh@beingmeta.com>  Sat, 05 Mar 2016 22:12:07 -0500

framerd (3.8.4-56) unstable; urgency=low

  * Fix bug in JWT changes
  * git patch version framerd-3.8.4-145-gebeeed4

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 10:43:32 -0500

framerd (3.8.4-55) unstable; urgency=low

  * Add RSA algorithm to JWT
  * Improve and fix parts of JWT module
  * Add third arg (urish) to ->base64 functions to use - and _
    for + and /
  * git patch version framerd-3.8.4-143-gdb7869b

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 08:57:17 -0500

framerd (3.8.4-54) unstable; urgency=low

  * Record boot handler and status in FramerD run dir
  * git patch version framerd-3.8.4-132-g22819cc (pre commit)

 -- Ken Haase <kh@beingmeta.com>  Mon, 29 Feb 2016 23:08:03 +0000

framerd (3.8.4-53) unstable; urgency=low

  * Fixes to daemon/servlet management scripts
  * git patch version framerd-3.8.4-128

 -- Ken Haase <kh@beingmeta.com>  Sun, 28 Feb 2016 20:09:55 +0000

framerd (3.8.4-52) unstable; urgency=low

  * Fix set-cookie to not automatically add dot to cookie domains
  * Fixes and refinements to jwt and jwt/auth
  * Make parsetime handle non-string arguments
  * git patch version framerd-3.8.4-126

 -- Ken Haase <kh@beingmeta.com>  Fri, 26 Feb 2016 22:58:38 +0000

framerd (3.8.4-51) unstable; urgency=low

  * Remove uridecode handling of '+' escapes
  * git patch version framer-3.8.4-121

 -- Ken Haase <kh@beingmeta.com>  Tue, 23 Feb 2016 18:51:41 +0000


src/exe/knox.c
/* -*- Mode: C; Character-encoding: utf-8; -*- */

/* Copyright (C) 2004-2020 beingmeta, inc.
   This file is part of beingmeta's Kno platform and is copyright
   and a valuable trade secret of beingmeta, inc.
*/

#ifndef EMBEDDED_KNO
#ifndef _FILEINFO
#define _FILEINFO __FILE__
#endif
#endif

#include "kno/knosource.h"
#include "kno/defines.h"
#include "kno/lisp.h"
#include "kno/support.h"
#include "kno/tables.h"
#include "kno/storage.h"
#include "kno/drivers.h"
#include "kno/eval.h"
#include "kno/ports.h"
#include "kno/cprims.h"

#include <libu8/libu8.h>
#include <libu8/u8timefns.h>
#include <libu8/u8pathfns.h>
#include <libu8/u8filefns.h>
#include <libu8/u8stringfns.h>
#include <libu8/u8fileio.h>
#include <libu8/u8stdio.h>
#include <libu8/u8printf.h>

#include <stdlib.h>
#include <stdio.h>
#include <locale.h>
#include <strings.h>
#include <sys/time.h>
#include <math.h>
#include <time.h>

#include "main.h"

#define MAX_CONFIGS 512

#include "main.c"

static u8_string stop_file=NULL, done_file=NULL;
static u8_string wait_for_file = NULL;

static int debug_maxelts = 32, debug_maxchars = 80;

static char *configs[MAX_CONFIGS], *exe_arg = NULL, *source_arg = NULL;
static int n_configs = 0, eval_stdin = 0;

static int chain_fast_exit=1;

static u8_condition FileWait=_("FILEWAIT");

static void exit_kno()
{
  if (!(kno_be_vewy_quiet)) kno_log_status("Exit(kno)");
}

static lispval load_stdin(kno_lexenv env)
{
  static U8_XINPUT u8stdin;
  u8_string cwd = u8_getcwd();
  u8_init_xinput(&u8stdin,0,NULL);
  lispval v = kno_load_stream((u8_input)&u8stdin,env,cwd);
  u8_close_input((u8_input)&u8stdin);
  u8_free(cwd);
  return v;
}

typedef char *charp;

static lispval main_symbol = KNO_VOID, exec_source = KNO_FALSE;
static lispval real_main = KNO_VOID;

static lispval chain_prim(int n,lispval *args)
{
  if (n_configs>=MAX_CONFIGS)
    return kno_err(_("Too many configs to CHAIN"),"chain_prim",NULL,VOID);
  else if ( (stop_file) && (u8_file_existsp(stop_file)) ) {
    u8_log(LOG_CRIT,"StopFile",
	   "Not chaining because the file '%s' exists",stop_file);
    return KNO_FALSE;}
  else if ( (done_file) && (u8_file_existsp(done_file)) ) {
    u8_log(LOG_CRIT,"DoneFile",
	   "Not chaining because the file '%s' exists",done_file);
    return KNO_FALSE;}
  else {
    int i = 0, cargc = 0, rv = -1;
    /* This stream will contain the chaining message */
    struct U8_OUTPUT argstring;
    char **cargv = u8_alloc_n(n+n_configs+4,charp);
    U8_INIT_STATIC_OUTPUT(argstring,512);
    cargv[cargc++]=exe_arg;
    cargv[cargc++]=source_arg;
    i = 0; while (i<n)
	     if (KNO_STRINGP(args[i])) {
	       u8_printf(&argstring," %s",CSTRING(args[i]));
	       cargv[cargc]=u8_tolibc(CSTRING(args[i]));
	       i++; cargc++;}
	     else {
	       u8_string as_string = kno_lisp2string(args[i]);
	       char *libc_string = u8_tolibc(as_string);
	       u8_printf(&argstring," %s",as_string);
	       u8_free(as_string);
	       cargv[cargc]=libc_string;
	       i++; cargc++;}
    u8_puts(&argstring," LOGAPPEND=yes");
    cargv[cargc++]=u8_strdup("LOGAPPEND=yes");
    i = 0; while (i<n_configs) {
      char *config = configs[i];
      if ( (strncmp(config,"LOGAPPEND=",10) == 0) ||
	   (strncmp(config,"PIDFILE=",10) == 0) )
	i++;
      else {
	u8_printf(&argstring," %s",u8_fromlibc(configs[i]));
	cargv[cargc++]=configs[i++];}}
    cargv[cargc++]=NULL;
    fflush(stdout); fflush(stderr);
    /* TODO: Should we run any exit methods here? */
    u8_log(LOG_INFO,"CHAIN","Closing pools and indexes");
    kno_fast_exit = chain_fast_exit;
    kno_close_pools();
    kno_close_indexes();
    u8_log(LOG_NOTICE,"CHAIN",">> %s %s%s",
	   exe_arg,u8_fromlibc(source_arg),argstring.u8_outbuf);
    u8_free(argstring.u8_outbuf);
    u8_log(LOG_NOTICE,"CHAIN","%s",
	   "================================================================"
	   "================================");
    rv = execvp(exe_arg,cargv);
    if (rv<0) {
      u8_graberr(errno,"CHAIN",u8_strdup(source_arg));
      return KNO_ERROR;}
    else return KNO_INT(rv);}
}

static void print_args(int argc,char **argv)
{
  int j = 0;
  fprintf(stderr,"-- argc=%d pid=%d\n",argc,getpid());
  while (j<argc) {
    fprintf(stderr,"   argv[%d]=%s\n",j,argv[j]);
    j++;}
}

static lispval *handle_argv(int argc,char **argv,size_t *arglenp,
			    u8_string *exe_namep,
			    u8_string *source_specp,
			    u8_string appid_prefix)
{
  lispval *args = NULL;
  u8_string tmp_string = NULL, source_spec = NULL, exe_name = NULL;
  /* Bit map of args which we handle */
  unsigned char arg_mask[argc];	 memset(arg_mask,0,argc);
  int i = 1;

  if (getenv("KNO_SHOWARGV")) print_args(argc,argv);

  exe_arg = argv[0];

  tmp_string = u8_fromlibc(exe_arg);
  exe_name = u8_basename(tmp_string,NULL);
  u8_free(tmp_string); tmp_string = NULL;
  if (exe_namep) *exe_namep = exe_name;

  while (i<argc) {
    if (isconfig(argv[i])) {
      u8_log(LOG_INFO,"Knoconfig","    %s",argv[i]);
      if (n_configs >= MAX_CONFIGS) n_configs++;
      else configs[n_configs++] = u8_strdup(argv[i]);
      i++;}
    else if (source_spec) {
      u8_log(LOG_INFO,"ARG","    %s",argv[i]);
      i++;}
    else {
      u8_log(LOG_INFO,"Source","    %s",argv[i]);
      source_arg = argv[i];
      source_spec = u8_fromlibc(source_arg);
      /* The arg_mask is used to tell kno_handle_argv which arguments
	 to leave out of the arguments passed to the script */
      arg_mask[i] = 'X';
      i++;}}

  if (source_spec == NULL) {}
  else if (source_specp) {
    *source_specp = source_spec;}
  else {}

  init_libraries();

  args = kno_handle_argv(argc,argv,arg_mask,arglenp);

  if (u8_appid() == NULL) {
    u8_string base;
    if (source_spec)
      base = u8_basename(source_spec,"*");
    else base = u8_basename(exe_name,NULL);
    if (appid_prefix == NULL)
      u8_default_appid(base);
    else {
      u8_string combined = u8_string_append(appid_prefix,base,NULL);
      u8_default_appid(combined);
      u8_free(combined);}
    u8_free(base);}

  if (source_specp == NULL) u8_free(source_spec);
  if (exe_namep == NULL) u8_free(exe_name);

  return args;
}

static void link_local_cprims()
{
}

int do_main(int argc,char **argv,
	    u8_string exe_name,u8_string source_spec,
	    lispval *args,size_t n_args)
{
  kno_lexenv env = kno_working_lexenv();
  lispval main_proc = VOID, result = VOID;
  int retval = 0, exec_module = 0;
  u8_string source_base = NULL;

  DEFC_PRIMN("CHAIN",chain_prim,KNO_VAR_ARGS|MIN_ARGS(0),
	    "Resets the current process to a fresh instance of knox");
  KNO_LINK_CPRIM("CHAIN",chain_prim,0,(lispval)env);

  link_local_cprims();

  u8_register_source_file(_FILEINFO);

  stop_file=u8_getenv("U8_STOPFILE");
  if (stop_file == NULL)
    stop_file=kno_runbase_filename(".stop");
  done_file=u8_getenv("U8_DONEFILE");
  if (done_file == NULL)
    done_file=kno_runbase_filename(".done");

  kno_set_app_env(env);

  main_symbol = real_main = kno_intern("main");

  u8_string main_ref =
    (source_spec == NULL) ? (NULL) :
    (source_spec[0]==':') ? (strchr(source_spec+1,':')) :
    (strchr(source_spec,':'));
  if (main_ref) {
    main_symbol = kno_getsym(main_ref+1);
    source_base = u8_slice(source_spec,main_ref);}
  else if (source_spec)
    source_base = u8_strdup(source_spec);
  else source_base=NULL;

  if (source_base)
    exec_source = kno_wrapstring(source_base);

  kno_register_config
    ("DEBUGMAXCHARS",
     _("Max number of chars in strings output in error reports"),
     kno_intconfig_get,kno_intconfig_set,
     &debug_maxchars);
  kno_register_config
    ("DEBUGMAXELTS",
     _("Max number of sequence/choice elements to display in error reports"),
     kno_intconfig_get,kno_intconfig_set,
     &debug_maxelts);
  kno_register_config
    ("FILEWAIT",
     _("File to wait to exist before starting"),
     kno_sconfig_get,kno_sconfig_set,
     &wait_for_file);
  kno_register_config
    ("STOPFILE",
     _("File to wait to exist before starting"),
     kno_sconfig_get,kno_sconfig_set,
     &stop_file);
  kno_register_config
    ("EVALSTDIN",_("Read additional source from STDIN"),
     kno_boolconfig_get,kno_boolconfig_set,
     &eval_stdin);
  kno_register_config
    ("MAIN",
     _("The name of the (main) routine for this file"),
     kno_lconfig_get,kno_symconfig_set,
     &main_symbol);
  kno_register_config
    ("EXECSOURCE",
     _("The name of the root file being executed for this application"),
     kno_lconfig_get,NULL,&exec_source);

  setlocale(LC_ALL,"");
  /* Process command line arguments */

  atexit(exit_kno);

  /* INITIALIZING MODULES */
  /* Normally, modules have initialization functions called when
     dynamically loaded.  However, if we are statically linked, or we
     don't have the "constructor attributes" use to declare init functions,
     we need to call some initializers explicitly. */

  if (!(kno_be_vewy_quiet)) kno_boot_message();
  if ( (stop_file) && (u8_file_existsp(stop_file)) ) {
    u8_log(LOG_CRIT,"StopFile",
	   "Not starting because the file '%s' exists",stop_file);
    return 1;}
  else if (wait_for_file) {
    if (u8_file_existsp(wait_for_file))
      u8_log(LOG_NOTICE,FileWait,"Starting now because '%s' exists",
	     wait_for_file);
    else {
      int n = 0;
      u8_log(LOG_NOTICE,FileWait,"Waiting for '%s' to exist",
	     wait_for_file);
      while (1) {
	n++; if (n<15) sleep(n); else sleep(15);
	if (u8_file_existsp(wait_for_file)) {
	  u8_log(LOG_NOTICE,FileWait,"[%d] Starting now because '%s' exists",
		 n,wait_for_file);
	  break;}
	else if ((n<15) ? ((n%4)==0) : ((n%20)==0))
	  u8_log(LOG_NOTICE,FileWait,"[%d] Waiting for '%s' to exist",
		 n,wait_for_file);}}}

  if (source_spec == NULL) {}
  else if (strcmp(source_spec,"-")==0)
    result = load_stdin(env);
  else if (source_spec[0] == ':') {
    lispval modname = kno_getsym(source_spec+1);
    result = kno_find_module(modname,0);
    if (KNO_VOIDP(result))
      result = kno_err("NoExecModule","do_main",source_spec,KNO_VOID);
    else exec_module=1;}
  else if (u8_file_existsp(source_spec) ) {
    lispval src = kno_wrapstring(u8_realpath(source_spec,NULL));
    kno_set_config("SOURCE",src);
    result = kno_load_source(source_spec,env,NULL);
    if (KNO_ABORTED(result)) {
      u8_exception ex = u8_current_exception;
      if (ex == NULL)
	u8_seterr("LoadFailed","knox/main()",u8_strdup(source_spec));}
    kno_decref(src);}
  else result = KNO_VOID;

  if (!(kno_be_vewy_quiet)) {
    double startup_time = u8_elapsed_time()-kno_load_start;
    char *units="s";
    if (startup_time>1) {}
    else if (startup_time>0.001) {
      startup_time = startup_time*1000; units="ms";}
    else {startup_time = startup_time*1000000; units="ms";}
    u8_message("Kno %s loaded in %0.3f%s, %d/%d pools/indexes",
	       u8_appid(),startup_time,units,kno_n_pools,
	       kno_n_primary_indexes);}

  if (KNO_ABORTED(result)) {}
  else if (!(SYMBOLP(main_symbol))) {
    kno_decref(result);
    result = kno_err("BadMain","knox.main()",source_spec,main_symbol);
    main_proc = KNO_VOID;}
  else if (exec_module) {
    if (KNO_LEXENVP(result)) {
      kno_lexenv env = (kno_lexenv) result;
      if (KNO_VOIDP(env->env_exports))
	main_proc = kno_get(env->env_bindings,main_symbol,KNO_VOID);
      else if (KNO_TABLEP(env->env_exports))
	main_proc = kno_get(env->env_exports,main_symbol,KNO_VOID);
      else main_proc = kno_get(env->env_bindings,main_symbol,KNO_VOID);}
    else main_proc = kno_get(result,main_symbol,KNO_VOID);
    if (KNO_VOIDP(main_proc)) {
      u8_log(LOG_ERROR,kno_NoHandler,"No handler for %s --- %q in %q",
	     source_spec,main_symbol,result);
      main_proc = kno_err(kno_NoHandler,"knox",source_spec,result);}
    else if (!(KNO_APPLICABLEP(main_proc))) {
      u8_log(LOG_ERROR,kno_BadHandler,"No handler for %s --- %q in %q\n  %q",
	     source_spec,main_symbol,result,main_proc);
      lispval err = kno_err(kno_BadHandler,"knox",source_spec,main_proc);
      kno_decref(main_proc);
      main_proc=err;}}
  else main_proc = kno_symeval(main_symbol,env);

  if (!(KNO_ABORTP(result))) {
    if (KNO_VOIDP(main_proc)) {
      if (source_spec == NULL) {
	int i = 0;
	if (main_symbol == real_main)
	  fprintf(stderr,"Error: %s no filename specified or (MAIN) defined",
		  exe_name);
	else fprintf(stderr,"Error: %s no filename specified or main (%s) defined",
		     exe_name,KNO_SYMBOL_NAME(main_symbol));
	while (i<argc) {
	  fprintf(stderr,"argv[%d]=%s\n",i,argv[i]);
	  i++;}
	fprintf(stderr,"Usage: %s filename [config = val]*\n",exe_name);
	result = kno_err("No MAIN","main()",NULL,main_symbol);}
      else {
	/* Nothing to do */
	kno_decref(result);
	result = KNO_VOID;}}
    else if (KNO_ABORTED(main_proc))
      result = kno_incref(main_proc);
    else if (KNO_APPLICABLEP(main_proc)) {
      kno_decref(result);
      result = kno_apply(main_proc,n_args,args);}
    else {
      u8_log(LOGWARN,"BadMain",
	     "The main procedure for %s (%q) isn't applicable",
	     ((source_spec) ? (source_spec) : (U8S("stdin")) ),
	     main_proc);
      result = kno_err("BadMain","main()",KNO_SYMBOL_NAME(main_symbol),
		       main_proc);}}
  if (source_spec) source_spec = NULL;
  if (KNO_TROUBLEP(result)) {
    u8_exception e = u8_erreify();
    if (e == NULL) {
      fputs("\nNull error object!\n",stderr);
      retval = -1;}
    else {
      // lispval handler = kno_symeval(kno_intern("onerror");
      // if (KNO_APPLICABLEP(handler)) {} else NO_ELSE;
      U8_OUTPUT out; U8_INIT_OUTPUT(&out,10000);
      kno_unparse_maxchars = debug_maxchars;
      kno_unparse_maxelts = debug_maxelts;
      kno_output_errstack(&out,e);
      fputs(out.u8_outbuf,stderr);
      fputc('\n',stderr);
      /* Write out the exception object somehow */
      u8_close_output(&out);
      u8_free_exception(e,1);
      retval = -1;}}

  if ( ! kno_fast_exit ) {
    kno_decref(result);
    kno_decref(main_proc);}

  /* Run registered thread cleanup handlers.
     Note that since the main thread wasn't started with a function which
     calls u8_threadexit(), we do it here. */
  u8_threadexit();

  return retval;
}

#ifndef EMBEDDED_KNO
int main(int argc,char **argv)
{
  u8_string source_spec = NULL, exe_name = NULL;
  lispval *args = NULL; size_t n_args = 0;
  int i = 0, retval = 0;

  u8_log_show_procinfo=1;

  kno_main_errno_ptr = &errno;

  KNO_INIT_STACK_ROOT();

#if KNO_TESTCONFIG
  kno_autoload_config("TESTMODS","TESTLOAD","TESTINIT");
#endif

  args = handle_argv(argc,argv,&n_args,&exe_name,&source_spec,NULL);

  retval = do_main(argc,argv,exe_name,source_spec,args,n_args);

  kno_pop_stack(_stack);

  i = 0; while (i<n_args) {
    lispval arg = args[i++];
    kno_decref(arg);}
  u8_free(args);

  if (exe_name) u8_free(exe_name);
  if (source_spec) u8_free(source_spec);

  /* Call this here, where it might be easier to debug, even
     though it's alos an atexit handler */
  _kno_finish_threads();

 return retval;
}
#endif

